java/ql/src/semmle/code/java/security/UnsafeDeserialization.qll

```diff
@@ -114,6 +182,180 @@ class SafeKryo extends DataFlow2::Configuration {
   }
 }
 
+/**
+ * Tracks flow from `enableDefaultTyping` calls to a subsequent Jackson deserialization method call.
+ */
+private class EnableJacksonDefaultTypingConfig extends DataFlow2::Configuration {
+  EnableJacksonDefaultTypingConfig() { this = "EnableJacksonDefaultTypingConfig" }
+
+  override predicate isSource(DataFlow::Node src) {
+    any(EnableJacksonDefaultTyping ma).getQualifier() = src.asExpr()
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof ObjectMapperReadSink }
+}
+
+/**
+ * Tracks flow from calls, which set a type validator, to a subsequent Jackson deserialization method call,
+ * including across builder method calls.
+ *
+ * Such a Jackson deserialization method call is safe because validation will likely prevent instantiating unexpected types.
+ */
+private class SafeObjectMapperConfig extends DataFlow2::Configuration {
+  SafeObjectMapperConfig() { this = "SafeObjectMapperConfig" }
+
+  override predicate isSource(DataFlow::Node src) {
+    src instanceof SetPolymorphicTypeValidatorSource
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof ObjectMapperReadSink }
+
+  /**
+   * Holds if `fromNode` to `toNode` is a dataflow step
+   * that configures or creates an `ObjectMapper` via a builder.
+   */
+  override predicate isAdditionalFlowStep(DataFlow::Node fromNode, DataFlow::Node toNode) {
+    exists(MethodAccess ma, Method m | m = ma.getMethod() |
+      m.getDeclaringType() instanceof MapperBuilder and
+      m.getReturnType()
+          .(RefType)
+          .hasQualifiedName("com.fasterxml.jackson.databind.json",
+            ["JsonMapper$Builder", "JsonMapper"]) and
```

Comment:
Bear in mind that dataflow steps must be type-preserving. The Builder is a MapperBuilder but JsonMapper itself is not, so this is likely only working in one of those cases.

Reply:
Right, the `JsonMapper.Builder` is a `MapperBuilder` but `JsonMapper` is not. This predicate is supposed to cover methods that returns either `JsonMapper.Builder` or `JsonMapper` even if these classes are not relateives. For example, the predicate is supposed to cover both `polymorphicTypeValidator()` and `build()` calls here:

```java
ObjectMapper mapper = JsonMapper.builder().polymorphicTypeValidator(ptv).build();
```

I can split this predicate to two if necessary.

Reply:
Actually never mind sorry, isAdditionalFlowStep actually always adds taint steps. By using `DataFlow::Configuration` you don't get the normal additional taint steps (e.g. reading from a tainted array produces a tainted element), but this particular step will propagate taint across building steps and ultimately to the built JsonMapper.